Add Paid Course Upsell

fjorgemota opened the pull request:
Fixes #5129

### Changes proposed in this Pull Request

* Add the Paid Course upsell screen;
* Move the title from the steps in the Editor Wizard to the interior of their content;
* Add a way to add an image to the right side of each step, as per the designs;

### Testing instructions

1. Checkout this branch;
2. Guarantee that Sensei Pro isn't installed; (by the way, remember to backup your Sensei Pro directory if you're going to delete it through WordPress...hehe);
3. Create a new course;
4. Click "Next" on the modal that appears to see the Paid Course upsell screen;
5. Verify that the "Get Sensei Pro" button works and opens a new tab on the Sensei LMS page (with some UTM tracking codes in the query string, too). Also, verify that the editor wizard goes to the next step (the Course Pattern step);
6. Create another new course;
7. Click on the "Continue with Sensei Free" button and verify that it DOES NOT open a new tab, and just goes to the Course Pattern step;

### Screenshot of the current version

![Screenshot of the Paid Course Upsell screen](https://user-images.githubusercontent.com/529864/169920896-561ae2c1-8cb3-4c34-805b-d453b6ac78f3.png)

### To do

- [x] Fix price font size;
- [x] Adjust font style of the text "per year, 1 site" just below the price;
- [x] Add a background image covering both the content area of the modal + the header (I accept ideas and suggestions here! Still struggling a little bit with the CSS part..hehe) (**WIP**);
- [x] Add analytics parameters to the link opened when clicking on the "Get Sensei Pro" button?

### Questions

- I accept suggestions of tests to do; =)
- I accept any feedback about the quality of the code;

fjorgemota committed:
Fixes spacing of the header title and close button on the wizard modal

fjorgemota committed:
Start working on the Paid Course upsell step

fjorgemota committed:
Add the image to include in the Paid Course upsell

fjorgemota committed:
Add WIP CSS implementation to have the background on the Paid Course upsell

fjorgemota committed:
Fix padding and add paragraph to the image container

So we can have aligned texts for better demonstration

renatho commented on assets/admin/editor-wizard/style.scss:
```diff
@@ -39,5 +41,33 @@
 	& .components-modal__content {
 		padding: 0;
+
+		& .components-modal-columns {
```
For the classes we're creating, it's interesting to create our own classes. One example for this case:

```css
.sensei-editor-wizard-modal {
    ...

    &__columns { // .sensei-editor-wizard-modal__column
	    display: grid;
	    grid-template-columns: 50% 50%;
    }

    &__upsell-features { // .sensei-editor-wizard-modal__upsell-features
	    list-style: none;
	    padding: 0;
    }

    &__upsell-feature-item { // .sensei-editor-wizard-modal__upsell-feature-item
        padding-left: 1.3em;

        &::before {
		    content: "\f00c";
		    font-family: FontAwesomeSensei, FontAwesome;
		    display: inline-block;
            ...
        }
    }
    ...
}
```

fjorgemota replied:
Fixed in [0da0194](https://github.com/Automattic/sensei/pull/5169/commits/0da019486eabdfce8415fefc584e60dc4b61f67a). Thanks for the tip!

renatho commented on assets/admin/editor-wizard/steps/upgrade-step.js:
```diff
@@ -1,25 +1,52 @@
+		<div className="components-modal-columns">
+			<div className="components-modal-columns__description">
+				<p>
+					Do you want to sell this course? This requires Sensei Pro
```
Just a reminder that we need to translate all texts using the i18n functions. 😉

fjorgemota replied:
Fixed in [0d40106](https://github.com/Automattic/sensei/pull/5169/commits/0d40106a5438878a3c707e402aa5ef9a71ae811a). Thanks for the reminder!

renatho commented on assets/admin/editor-wizard/steps/upgrade-step.js:
```diff
@@ -1,25 +1,52 @@
 };
 
-UpgradeStep.Title = 'Upgrade Step';
+UpgradeStep.Title = 'Sell with Sensei Pro';
```
I'd recommend as part of this PR a refactor to the title. We can just not send a title to the `<Modal>`, and add the title as part of the content in the steps. Here's an example that we have of a modal without a title in Gutenberg: https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/guide/index.js

fjorgemota replied:
Done in the commits [e470151](https://github.com/Automattic/sensei/pull/5169/commits/e470151863acfd8d69b12cbcf7be6c5a2a01a1db) and [bff4106](https://github.com/Automattic/sensei/pull/5169/commits/bff4106c9fb5fe5b23a1c89389c5a68abaca144a) (and I added you as a co-author after our pair programming session, hehe)

fjorgemota committed:
Refactor list of upsell features

As suggested by Renatho.

fjorgemota committed:
Refactor styles of the editor wizard

So we can remove margins of the modal and remove the title from the header.

Co-authored-by: Renatho De Carli Rosa <renatho@gmail.com>

fjorgemota committed:
Move Title to the interior of the Step

So we can customize it easily.

Co-authored-by: Renatho De Carli Rosa <renatho@gmail.com>

fjorgemota committed:
Fix typo on the upgrade step of the Editor Wizard

fjorgemota committed:
Add image to the Paid Course Upsell in the Editor Wizard

fjorgemota committed:
Delete test related to onStepChange, which was removed in the commit bff4106c9fb5fe5b23a1c89389c5a68abaca144a

fjorgemota committed:
Separate upsell classes into its own namespace

And fix pricing and pricing detail font sizes

fjorgemota committed:
Add images of the paid course upsell as placeholders

So we can least look at how other editor wizard steps will behave. =)

fjorgemota committed:
Add link to Sensei Pricing page with UTM tracking code

fjorgemota committed:
Rename Upgrade Step to Course Upgrade Step

So we can know that it applies only for the Course editor wizard easily.

fjorgemota committed:
Add calls to i18n functions in paid course upsell in editor wizard

So we can have these strings translated properly :D

fjorgemota committed:
Add hook to get details about the Sensei Pro extension

fjorgemota committed:
Hide Paid Course Upsell if the Sensei Pro plugin is installed

fjorgemota committed:
Add dynamic price to Paid Course Upsell

So we don't need to go looking around for updating the price in the code..:)

fjorgemota commented on assets/admin/editor-wizard/use-sensei-pro-extension.js:
```diff
@@ -0,0 +1,23 @@
+ */
+import { EXTENSIONS_STORE } from '../../extensions/store';
+
+export default function useSenseiProExtension() {
```
I created this hook based on [this component](https://github.com/Automattic/sensei/blob/a21c5162e31bc88079a76301fe6740aae2b6699d/assets/extensions/featured-product-sensei-pro.js) that I found inside the Sensei Core repository. I accept any feedbacks about how to improve this, as I need to get the extensions detail both on the Wizard component (to remove the Paid Course upsell if Sensei Pro extension is installed) as well as on the Course Upgrade Step (to get the price dynamically).

fjorgemota replied:
Moved the discussion to [this thread](https://github.com/Automattic/sensei/pull/5169#pullrequestreview-983182993) instead (after moving the hook out of the editor-wizard folder)

fjorgemota committed:
Move useSenseiProExtension to avoid code duplication

So we can re-use the logic in the FeaturedProductSenseiPro component too! :D

fjorgemota committed:
Fix image path in the plugin directory

Because I just observed that the image wasn't working in the final plugin build

fjorgemota commented on assets/extensions/use-sensei-pro-extension.js:
```diff
@@ -0,0 +1,23 @@
+ */
+import { EXTENSIONS_STORE } from './store';
+
+export default function useSenseiProExtension() {
```
I created this hook based on [this component](https://github.com/Automattic/sensei/blob/a21c5162e31bc88079a76301fe6740aae2b6699d/assets/extensions/featured-product-sensei-pro.js) that I found inside the Sensei Core repository. I accept any feedback about how to improve this, as I need to get the extension's detail both on the Wizard component (to remove the Paid Course upsell if Sensei Pro extension is installed) as well as on the Course Upgrade Step (to get the price dynamically).

renatho replied:
I was thinking that maybe another selector in the store to get the Sensei Pro extension would fit better for this purpose. But I think it also works well.

fjorgemota replied:
Changed in [9635974](https://github.com/Automattic/sensei/pull/5169/commits/9635974f54dd32a84ab5a5251adbaa62be67d7d9). Thanks!

renatho commented on assets/admin/editor-wizard/editor-wizard-modal.js:
```diff
@@ -94,12 +94,19 @@ const EditorWizardModal = () => {
+	const senseiProExtension = useSenseiProExtension();
+
+	if ( ! senseiProExtension || senseiProExtension.is_installed === true ) {
+		stepsByPostType.course = [ CourseDetailsStep, CoursePatternsStep ];
```
Would it be better if we use an array filter instead of creating the array again? Just to avoid this duplication of the other steps. So we avoid issues in the future while maintaining it. :)

fjorgemota replied:
Adjusted in [c5c0094](https://github.com/Automattic/sensei/pull/5169/commits/c5c00940cbe91ff9ddc3d10b559f064b4f15f251). Thanks!

renatho commented on assets/admin/editor-wizard/steps/course-upgrade-step.js:
```diff
@@ -0,0 +1,95 @@
+ */
+import { Button } from '@wordpress/components';
+import { __, sprintf } from '@wordpress/i18n';
+import useSenseiProExtension from '../../../extensions/use-sensei-pro-extension';
```
We should add another comment here to Internal dependencies.

fjorgemota replied:
Fixed in [c5e62a7](https://github.com/Automattic/sensei/pull/5169/commits/c5e62a7ca289f1f11b309d4af7e11da4a7f231e8). Thanks!

renatho commented on assets/admin/editor-wizard/steps/course-upgrade-step.js:
```diff
@@ -0,0 +1,95 @@
+						'Illustration of a course listing with the pricing defined and with the button "Purchase Button"',
+						'sensei-lms'
+					) }
+					height="75%"
```
I think we could set the site in the CSS in order to keep all the styles there.

fjorgemota replied:
I thought about doing that, but at the same time, I feel we **could** have images of different sizes, which may need different values for height.

WDYT?

renatho replied:
I personally would prefer to have all the styles in the CSS unless it would be really necessary to be something dynamic. But I think that's fine here. 😉

fjorgemota replied:
Fixed in [d962689](https://github.com/Automattic/sensei/pull/5169/commits/d96268987e23bd157ed79e8fbe877e87e91dc36e). Thanks! :)

renatho reviewed (commented):
This is looking great! I just added some small suggestions.

> I accept suggestions of tests to do; =)

Just in order to merge it earlier, I'd do it in another PR. I'd test the following parts:

- `useSenseiProExtension` - We could mock the `useSelect`, and confirm if it's filtering the Sensei Pro properly.
- `EditorWizardModal` - It'd be interesting to check if the upsell step is added when should and not when shouldn't. I think the easier way would be mocking the `Wizard` component, and checking if CourseUpgradeStep is in the array or not.
- `CourseUpgradeStep` - We could also check if the price is rendered correctly and if the actions work properly.

renatho commented:
Just one other thing while checking the design:

- The features list should have a `font-weight: 700`.
- _"per year, 1 site"_ should have a different spacing above and below the text.
- I think it's not part of this PR, but if you could, we could take the opportunity to fix the spacing between the buttons in the `Actions`. It should have a `12px` spacing between them. And the padding in the whole Action container should be `18px 30px` instead of `30px`.

fjorgemota committed:
Remove CourseUpgradeStep from stepsByPostType.course using filter

Instead of copying the array again without the step.

fjorgemota committed:
Add comment defining Internal Dependencies for Course Upgrade Step

fjorgemota committed:
Remove useSenseiProExtension in favour of a selector

I had to add a resolver in the store to be able to load the extensions and so to identify the extension appropriately.

fjorgemota committed:
Fix button spacing on the Course Upgrade Step

fjorgemota committed:
Adjust price detail margin on the Course Upgrade Step

fjorgemota committed:
Fix font-weight on the list of features in the Course Upgrade Step

renatho commented on assets/admin/editor-wizard/editor-wizard-modal.js:
```diff
@@ -97,12 +97,11 @@ const EditorWizardModal = () => {
+	const { postType, senseiProExtension } = useSelect( ( select ) => ( {
 		postType: select( editorStore )?.getCurrentPostType(),
+		senseiProExtension: select( EXTENSIONS_STORE ).getSenseiProExtension(),
 	} ) );
```
Sorry that I just noticed now, but one thing here is that we just need to get it one time, so we can use the [`deps`](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-data/#useselect) as `[]`.

fjorgemota replied:
Fixed in [c2f7362](https://github.com/Automattic/sensei/pull/5169/commits/c2f736285a54f1671bd5786855d4a76e2c3146c5). Thanks!

fjorgemota committed:
Remove resolver in favour of using createRegistrySelector

fjorgemota committed:
Use CSS to set the height of the image instead of relying on img attributes

fjorgemota committed:
Pass [] as dependencies and use arrow functions on useSelect calls

renatho commented on assets/admin/editor-wizard/steps/lesson-details-step.js:
```diff
@@ -4,14 +4,25 @@
+						'assets/dist/images/sensei-pro-upsell.png'
+					}
+					alt="PENDING TO IMPLEMENT, BUT HERE TO SHOW IT WORKING"
+					className="sensei-editor-wizard-modal-upsell__image"
```
I think we should use a more generic name here, right?
Maybe `sensei-editor-wizard-modal__illustration-image`?

fjorgemota replied:
Changed in [fa0760a](https://github.com/Automattic/sensei/pull/5169/commits/fa0760a72c368db4a30903876b0d271f334a509a). Thanks!

renatho reviewed (dismissed):
This is very good! Just one final thing.
After that, since it's pretty simple, I think you can merge directly.

fjorgemota committed:
Rename CSS class of the images on the editor wizard as per suggestion

renatho reviewed (approved):
🎉🎉🎉
Great PR!

aaronfc commented on assets/admin/editor-wizard/editor-wizard-modal.js:
```diff
@@ -94,12 +94,25 @@ const EditorWizardModal = () => {
+		[]
+	);
+
+	if ( ! senseiProExtension || senseiProExtension.is_installed === true ) {
+		stepsByPostType.course = stepsByPostType.course.filter(
+			( step ) => step !== CourseUpgradeStep
+		);
+	}
```
This is making the Upgrade step flash in/out if you go over the first step somewhat fast.
https://user-images.githubusercontent.com/799065/170244529-f19593a5-145b-42e3-baa5-b493d77157c7.mp4

Any idea on how to workaround that?

aaronfc reviewed (commented):
Sorry for the late review! It is looking nice but I found some "flashing" behaviour I reported in a comment.
Maybe we want to look into it.
